File: src/topols/embedding/mcts.py
"""Anytime Monte Carlo Tree Search over `EmbeddingState`s: UCT selection,
greedy rollouts, and a best-so-far result within an iteration and
wall-clock budget.
"""

import math
import time

from topols.routing.astar import WORK, WORK_PER_SECOND

# The wall clock is only a safety net: a call may run this many times its
# budget in real seconds before it is stopped regardless of work done.
SAFETY_FACTOR = 10

# ---------------------------------------------------------------------------
# Optional diagnostics: a caller may set STATS_SINK to a list to record, per
# mcts() call, how many iterations completed and whether the wall-clock budget
# or the iteration cap ended the search. None (the default) records nothing.
STATS_SINK = None

# Separate opt-in sink for the reward-cache hit-rate diagnostic below.
CACHE_STATS_SINK = None

# In-loop reward cache: when selection lands on a terminal node whose reward
# is already known, reuse it instead of rerunning rollout()/reward(). A module
# toggle so it can be switched off for comparisons.
ENABLE_INLOOP_REWARD_CACHE = True

# Sentinel for "no cached reward yet". Distinct from None because reward()
# legitimately returns None when a terminal state's ceiling or T-gate routing
# fails, and those states must stay cacheable.
_UNSET = object()

# ---------------------------------------------------------------------------
# MCTS node in the search tree
# ---------------------------------------------------------------------------

class MCTSNode:
    """A node of the search tree.

    Attributes:
        state: the `EmbeddingState` at this node.
        parent, children: tree links.
        visits, value: visit count and summed backed-up reward (UCT statistics).
        untried: moves of `state` not expanded yet (`EmbeddingState.moves`).
        cached_reward: reward of `state` once known (terminal states only);
            `_UNSET` until then.
    """
    __slots__ = ("state","parent","children",
                 "visits","value","untried","try_flag", "id", "cached_reward")

    def __init__(self, state, parent=None, move_num=None, block_switch=False, ceiling_switch=False):
        self.state = state
        self.parent = parent
        self.children = []
        self.visits = 0
        self.value = 0.0
        self.untried = state.moves(num=move_num, block_switch=block_switch, ceiling_switch=ceiling_switch)
        self.cached_reward = _UNSET

    def uct_select_child(self, c=0.7):
        """Child with the highest UCB1 score, `mean value + c * sqrt(ln N / n)`."""
        best = None
        best_ucb = -1e9
        for child in self.children:
            ucb = (child.value/child.visits +
                   c * math.sqrt(math.log(self.visits) / child.visits))
            if ucb >= best_ucb:
                best_ucb = ucb
                best = child
        return best


# ---------------------------------------------------------------------------
# Roll‑out policy   (greedy placement to nearest free voxel)
# ---------------------------------------------------------------------------

def rollout(state, max_steps=2000, obj=None, layer=None, block_switch=False, ceiling_switch=False, length=4):
    """Complete `state` greedily: at every step take the move whose
    resulting state has the smallest bounding-box volume.

    Returns:
        `(reward, terminal_state)` when the layer is completed and
        `EmbeddingState.reward` succeeds; the scalar `-1e9` otherwise
        (no legal move, a routing failure, or `max_steps` exhausted).
    """
    cur = state

    for j in range(max_steps):
        if cur.is_terminal():
            result = cur.reward(layer=layer, length=length)
            if result is None:
                return -1e9
            reward, _, _, _ = result
            return reward, cur

        moves = cur.moves(num=6, block_switch=block_switch, ceiling_switch=ceiling_switch, rollout=True)

        if not moves:
            return -1e9

        # choose the move giving smallest immediate bbox growth
        best_move = None
        best_vol = 1e9

        for m in moves:
            nxt = cur.next_state(m)
            if nxt is None:
                continue
            if nxt.vol < best_vol:
                best_vol = nxt.vol
                best_move = m

        if best_move is None:
            return -1e9

        cur = cur.next_state(best_move)
        if cur is None:
            return -1e9

    return -1e9  # safeguard


# ---------------------------------------------------------------------------
# Main MCTS loop
# ---------------------------------------------------------------------------

def mcts(root_state, iters=10000, time_limit=None, obj=None, move_num=None, block_switch=False, ceiling_switch=False, layer=None, length=None):
    """Search for the lowest-volume complete embedding of one layer.

    Standard UCT loop (select, expand one child, greedy rollout, back up)
    over `EmbeddingState`s, stopped by whichever of `iters` iterations or
    the work budget comes first. The budget is `time_limit` seconds of
    search work on the reference machine, measured in A* expansions
    (`routing.astar.WORK`), so a call does the same work -- and returns
    the same result -- on any machine; the real clock only acts as a
    safety net. The search is *anytime*: for a fixed random state and
    `root_state` the iteration sequence is deterministic, and the result
    is the best complete state seen anywhere (in the tree or in a
    rollout), so a larger budget never returns a worse state.

    Args:
        root_state: `EmbeddingState` with the previous layer embedded and
            nothing of the current layer placed yet.
        iters: maximum number of iterations.
        time_limit: work budget, in seconds of reference-machine search
            (None = unlimited).
        move_num: number of candidate placements generated per node
            (`EmbeddingState.moves(num=...)`).
        block_switch, ceiling_switch: passed to `moves`; the first block
            of a compile and a search started from a lifted ceiling allow
            different placements.
        layer, length: forwarded to `EmbeddingState.reward`.

    Returns:
        The best terminal `EmbeddingState`, or None if no rollout completed
        the layer.
    """
    root = MCTSNode(root_state, move_num=move_num, block_switch=block_switch, ceiling_switch=ceiling_switch)
    work_budget = time_limit * WORK_PER_SECOND if time_limit else float("inf")
    work_start = WORK[0]
    deadline = time.time() + (SAFETY_FACTOR * time_limit if time_limit else 1e9)

    best_rollout = -1e9
    best_rollout_state = None

    for i in range(iters):
        if WORK[0] - work_start >= work_budget or time.time() > deadline:
            if STATS_SINK is not None:
                STATS_SINK.append({"iters_completed": i, "iters_requested": iters, "search_bound": True, "layer": layer})
            break

        node = root

        # 1. Selection
        while not node.untried and node.children:
            node = node.uct_select_child()

        # 2. Expansion
        if node.untried:
            move = node.untried.pop()
            nxt_state = node.state.next_state(move)
            if nxt_state is None:
                continue
            node = MCTSNode(nxt_state, parent=node, move_num=move_num, block_switch=block_switch, ceiling_switch=ceiling_switch)
            node.parent.children.append(node)

        # 3. Simulation
        if (ENABLE_INLOOP_REWARD_CACHE
                and node.cached_reward is not _UNSET
                and node.state.is_terminal()):
            # Selection reached a terminal node whose reward is already cached;
            # rollout() would recompute exactly this value.
            reward = node.cached_reward
            rollout_state = node.state
            if CACHE_STATS_SINK is not None:
                CACHE_STATS_SINK.append({"cache_hit": True, "layer": layer})
        else:
            reward = rollout(node.state, layer=layer, block_switch=block_switch, ceiling_switch=ceiling_switch, length=length)
            if reward != -1e9:
                reward, rollout_state = reward
                # Cache the reward for the post-loop retrieval and later revisits.
                if node.state is rollout_state:
                    node.cached_reward = reward
            if CACHE_STATS_SINK is not None and node.state.is_terminal():
                CACHE_STATS_SINK.append({"cache_hit": False, "layer": layer})
        if reward != -1e9:
            if reward > best_rollout:
                best_rollout = reward
                best_rollout_state = rollout_state

        # 4. Back-propagation
        while node:
            node.visits += 1
            node.value += reward
            node = node.parent
    else:
        # Loop completed without `break` -- iters-bound, not search-bound.
        if STATS_SINK is not None:
            STATS_SINK.append({"iters_completed": iters, "iters_requested": iters, "search_bound": False, "layer": layer})

    # retrieve best completed embedding seen
    best_state = None
    best_val = -1e9
    stack = [root]
    while stack:
        n = stack.pop()
        if n.state.is_terminal():
            # Reuse the cached reward (reward() does real routing work).
            if n.cached_reward is not _UNSET:
                r_val = n.cached_reward
            else:
                r = n.state.reward(length=length)
                r_val = r[0] if r is not None else None
            if r_val is not None and r_val > best_val:
                best_val = r_val
                best_state = n.state
        stack.extend(n.children)

    # Return the best state seen anywhere. The tree only holds the terminals
    # that happened to be expanded (the first to appear is often poor), while
    # best_rollout_state is the best over every rollout; taking the better of
    # the two makes the result the best-so-far of a deterministic iteration
    # sequence, so a longer budget can never return a worse state.
    if best_state is None or (best_rollout_state is not None and best_rollout > best_val):
        best_state = best_rollout_state

    return best_state

File: rust/topols_core/src/embedding/mcts.rs
//! Anytime MCTS over `EmbeddingState`s (mcts.py): UCT selection, greedy
//! rollouts, best-so-far result within a work budget.

use std::rc::Rc;
use std::time::Instant;

use crate::embedding::state::EmbeddingState;
use crate::pyrandom::PyRandom;
use crate::routing::astar::{work, WORK_PER_SECOND};

/// The wall clock is only a safety net (`mcts.SAFETY_FACTOR`).
pub const SAFETY_FACTOR: f64 = 10.0;
pub const FAIL: f64 = -1e9;

struct Node {
    state: Rc<EmbeddingState>,
    parent: Option<usize>,
    children: Vec<usize>,
    visits: u64,
    value: f64,
    untried: Vec<crate::geometry::Cell>,
    /// `cached_reward`: Some(Some(r)) known reward, Some(None) known failure
    /// (reward() returned None), None not cached.
    cached: Option<Option<f64>>,
}

pub struct SearchParams {
    pub iters: usize,
    /// work budget in seconds of reference-machine search; None = unlimited
    pub time_limit: Option<f64>,
    pub move_num: Option<usize>,
    pub block_switch: bool,
    pub ceiling_switch: bool,
    pub length: usize,
}

fn new_node(arena: &mut Vec<Node>, state: Rc<EmbeddingState>, parent: Option<usize>, rng: &mut PyRandom, p: &SearchParams) -> usize {
    let untried = state.moves(rng, p.move_num.unwrap_or(6), p.block_switch, p.ceiling_switch, false);
    arena.push(Node { state, parent, children: vec![], visits: 0, value: 0.0, untried, cached: None });
    arena.len() - 1
}

/// `uct_select_child` with c = 0.7; ties go to the later child (`>=`).
fn uct_select(arena: &[Node], n: usize) -> usize {
    let c = 0.7f64;
    let mut best = None;
    let mut best_ucb = -1e9f64;
    let ln_n = (arena[n].visits as f64).ln();
    for &ch in &arena[n].children {
        let child = &arena[ch];
        let ucb = child.value / child.visits as f64 + c * (ln_n / child.visits as f64).sqrt();
        if ucb >= best_ucb {
            best_ucb = ucb;
            best = Some(ch);
        }
    }
    best.unwrap()
}

pub enum Rollout {
    Fail,
    /// reward, terminal state, whether that state is the rollout's input itself
    Ok(f64, Rc<EmbeddingState>, bool),
}

/// `rollout`: greedy completion by smallest bounding-box volume.
pub fn rollout(state: Rc<EmbeddingState>, rng: &mut PyRandom, p: &SearchParams) -> Rollout {
    let mut cur = state;
    let mut same = true;
    for _ in 0..2000 {
        if cur.is_terminal() {
            return match cur.reward(p.length) {
                None => Rollout::Fail,
                Some(r) => Rollout::Ok(r.reward, cur, same),
            };
        }
        let moves = cur.moves(rng, 6, p.block_switch, p.ceiling_switch, true);
        if moves.is_empty() {
            return Rollout::Fail;
        }
        let mut best_move = None;
        let mut best_vol = 1e9f64;
        for &m in &moves {
            if let Some(nxt) = cur.next_state(m) {
                if nxt.vol < best_vol {
                    best_vol = nxt.vol;
                    best_move = Some(m);
                }
            }
        }
        let Some(bm) = best_move else { return Rollout::Fail };
        // Python recomputes the chosen successor (and so does its work count).
        match cur.next_state(bm) {
            None => return Rollout::Fail,
            Some(n) => cur = Rc::new(n),
        }
        same = false;
    }
    Rollout::Fail
}

/// `mcts`: best terminal state found within the budget, or None.
pub fn mcts(root_state: Rc<EmbeddingState>, rng: &mut PyRandom, p: &SearchParams) -> Option<Rc<EmbeddingState>> {
    let mut arena: Vec<Node> = Vec::new();
    let root = new_node(&mut arena, root_state, None, rng, p);
    let work_budget = p.time_limit.map_or(f64::INFINITY, |t| t * WORK_PER_SECOND as f64);
    let work_start = work();
    let started = Instant::now();
    let deadline = p.time_limit.map_or(f64::INFINITY, |t| SAFETY_FACTOR * t);

    let mut best_rollout = FAIL;
    let mut best_rollout_state: Option<Rc<EmbeddingState>> = None;

    for _ in 0..p.iters {
        if (work() - work_start) as f64 >= work_budget || started.elapsed().as_secs_f64() > deadline {
            break;
        }
        // 1. selection
        let mut n = root;
        while arena[n].untried.is_empty() && !arena[n].children.is_empty() {
            n = uct_select(&arena, n);
        }
        // 2. expansion
        if let Some(mv) = arena[n].untried.pop() {
            let Some(nxt) = arena[n].state.next_state(mv) else { continue };
            let child = new_node(&mut arena, Rc::new(nxt), Some(n), rng, p);
            arena[n].children.push(child);
            n = child;
        }
        // 3. simulation
        let reward;
        let mut rollout_state: Option<Rc<EmbeddingState>> = None;
        if let (Some(Some(r)), true) = (arena[n].cached, arena[n].state.is_terminal()) {
            reward = r;
            rollout_state = Some(arena[n].state.clone());
        } else {
            match rollout(arena[n].state.clone(), rng, p) {
                Rollout::Fail => reward = FAIL,
                Rollout::Ok(r, st, same) => {
                    reward = r;
                    if same {
                        arena[n].cached = Some(Some(r));
                    }
                    rollout_state = Some(st);
                }
            }
        }
        if reward != FAIL && reward > best_rollout {
            best_rollout = reward;
            best_rollout_state = rollout_state;
        }
        // 4. back-propagation
        let mut m = Some(n);
        while let Some(i) = m {
            arena[i].visits += 1;
            arena[i].value += reward;
            m = arena[i].parent;
        }
    }

    // best completed embedding in the tree (DFS, stack order as in Python)
    let mut best_state: Option<Rc<EmbeddingState>> = None;
    let mut best_val = FAIL;
    let mut stack = vec![root];
    while let Some(i) = stack.pop() {
        if arena[i].state.is_terminal() {
            let r_val = match arena[i].cached {
                Some(c) => c,
                None => arena[i].state.reward(p.length).map(|r| r.reward),
            };
            if let Some(v) = r_val {
                if v > best_val {
                    best_val = v;
                    best_state = Some(arena[i].state.clone());
                }
            }
        }
        stack.extend(arena[i].children.iter().copied());
    }
    if best_state.is_none() || (best_rollout_state.is_some() && best_rollout > best_val) {
        best_state = best_rollout_state;
    }
    best_state
}
